apps/src/feedback.js

```diff
@@ -179,42 +177,70 @@ FeedbackUtils.prototype.displayFeedback = function(options, requiredBlocks,
   // be shown (including any feedback blocks), and add code to restore the
   // hint if the button gets pressed.
   if (hintRequestButton) {
-    // Swap out the specific feedback message with a generic one.
-    var genericFeedback = this.getFeedbackMessage_({message: msg.genericFeedback()});
-    var parentNode = feedbackMessage.parentNode;
-    parentNode.replaceChild(genericFeedback, feedbackMessage);
-
-    // If there are feedback blocks, temporarily remove them.
-    // Get pointers to the parent and next sibling so we can re-insert
-    // the feedback blocks into the correct location if needed.
-    var feedbackBlocksParent = null;
-    var feedbackBlocksNextSib = null;
-    if (feedbackBlocks && feedbackBlocks.div) {
-      feedbackBlocksParent = feedbackBlocks.div.parentNode;
-      feedbackBlocksNextSib = feedbackBlocks.div.nextSibling;
-      feedbackBlocksParent.removeChild(feedbackBlocks.div);
-    }
 
-    // If the user requests the hint...
-    dom.addClickTouchEvent(hintRequestButton, function() {
-      // Swap the specific feedback message back in.
-      parentNode.replaceChild(feedbackMessage, genericFeedback);
+    var already_seen = options.response &&
+        options.response.hint_view_requests &&
+        options.response.hint_view_requests.some(function (request) {
+          var matches = request.feedback_type === options.feedbackType;
```

Comment:
super mega tiny nit: variable name `matches` can read as either "this is a list of matches" or "did find a match". Something like `doesMatch` or `matchesSeenHint` could disambiguate.